packages/monitoring-system/data-processor/src/data-processors/cloud-logs-data-processor.ts

```diff
@@ -48,7 +88,341 @@ export class CloudLogsProcessor extends DataProcessor {
     this.listenLimit = options.listenLimit;
   }
 
+  /**
+   * Start the collection and processing task
+   *
+   * @returns a promise that settles once all received messages
+   * have been processed (successfully or unsuccessfully). The promise
+   * will be fulfilled even if some messages are unsuccessfully processed.
+   * The promise will only reject for runtime errors such as loss of
+   * communication to PubSub.
+   */
   public async collectAndProcess(): Promise<void> {
-    throw new Error('Method not implemented.');
+    this.subscription.on('message', this.processMessage.bind(this));
+    this.logger.debug('Now listening for PubSub messages');
+
+    return new Promise((resolve, reject) => {
+      const stopListening = () => {
+        this.subscription.removeAllListeners();
+        this.logger.debug(`Stopped listening after ${this.listenLimit}s`);
+
+        Promise.all(this.tasksInProgress)
+          .then(() => {
+            this.logger.debug('All messages processed');
+            resolve();
+          })
+          .catch(error => {
+            this.logger.error('Some messages could not be processed');
+            reject(error);
+          });
+      };
+
+      setTimeout(stopListening, this.listenLimit * 1000);
```

Comment:
nit: can you add comments about why this method is using a setTimeout

Reply:
will do!